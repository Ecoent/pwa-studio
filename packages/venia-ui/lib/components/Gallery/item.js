import React from 'react';
import { string, number, shape } from 'prop-types';
import { Link, resourceUrl } from '@magento/venia-drivers';
import { Price } from '@magento/peregrine';
import { transparentPlaceholder } from '@magento/peregrine/lib/util/images';

import { mergeClasses } from '../../classify';
import Image from '../Image';
import defaultClasses from './item.css';

// The placeholder image is 4:5, so we should make sure to size our product
// images appropriately.
const imageWidth = '300';
const imageHeight = '375';

const ItemPlaceholder = ({ classes }) => (
    <div className={classes.root_pending}>
        <div className={classes.images_pending}>
            <Image
                alt="Placeholder for gallery item image"
                classes={{ root: classes.image_pending }}
                src={transparentPlaceholder}
            />
        </div>
        <div className={classes.name_pending} />
        <div className={classes.price_pending} />
    </div>
);

// TODO: get productUrlSuffix from graphql when it is ready
const productUrlSuffix = '.html';

const GalleryItem = props => {
    const { item } = props;

    const classes = mergeClasses(defaultClasses, props.classes);

    if (!item) {
        return <ItemPlaceholder classes={classes} />;
    }

    const { name, price, small_image, url_key } = item;
    const productLink = `/${url_key}${productUrlSuffix}`;
    // See gallery.css for breakpoints and number of columns.
    const galleryBreakpoint = '640px';
    const sizeWhenTwoColumns = '50vw';
    const sizeWhenThreeColumns = '33vw'; // PR writeup: you can see this in action Responsive: 646 wide
    const sizes = `(max-width: ${galleryBreakpoint}) ${sizeWhenTwoColumns},
                   ${sizeWhenThreeColumns}`;

    return (
        <div className={classes.root}>
            <Link to={resourceUrl(productLink)} className={classes.images}>
                <Image
                    alt={name}
                    classes={{ root: classes.image }}
                    resource={small_image}
                    resourceHeight={imageHeight}
                    resourceWidth={imageWidth}
                    sizes={sizes}
                />
            </Link>
            <Link to={resourceUrl(productLink)} className={classes.name}>
                <span>{name}</span>
            </Link>
            <div className={classes.price}>
                <Price
                    value={price.regularPrice.amount.value}
                    currencyCode={price.regularPrice.amount.currency}
                />
            </div>
        </div>
    );
};

GalleryItem.propTypes = {
    classes: shape({
        image: string,
        image_pending: string,
        images: string,
        images_pending: string,
        name: string,
        name_pending: string,
        price: string,
        price_pending: string,
        root: string,
        root_pending: string
    }),
    item: shape({
        id: number.isRequired,
        name: string.isRequired,
        small_image: string.isRequired,
        url_key: string.isRequired,
        price: shape({
            regularPrice: shape({
                amount: shape({
                    value: number.isRequired,
                    currency: string.isRequired
                }).isRequired
            }).isRequired
        }).isRequired
    })
};

export default GalleryItem;
