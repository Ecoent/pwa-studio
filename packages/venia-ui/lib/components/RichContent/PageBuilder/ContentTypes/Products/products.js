import React, { useCallback } from 'react';
import { Query } from '@magento/venia-drivers';
import defaultClasses from './products.css';
import { mergeClasses } from '../../../../../classify';
import { arrayOf, shape, string } from 'prop-types';
import Gallery from '../../../../Gallery';
import getProductsBySku from '../../../../../queries/getProductsBySku.graphql';

/**
 * Page Builder Products component.
 *
 * This component is part of the Page Builder / PWA integration. It can be consumed without Page Builder.
 *
 * @typedef Products
 * @kind functional component
 *
 * @param {props} props React component props
 *
 * @returns {React.Element} A React component that displays a Products based on a number of skus.
 */
const Products = props => {
    const classes = mergeClasses(defaultClasses, props.classes);
    const {
        skus,
        textAlign,
        border,
        borderColor,
        borderWidth,
        borderRadius,
        marginTop,
        marginRight,
        marginBottom,
        marginLeft,
        paddingTop,
        paddingRight,
        paddingBottom,
        paddingLeft,
        cssClasses = []
    } = props;

    const dynamicStyles = {
        textAlign,
        border,
        borderColor,
        borderWidth,
        borderRadius,
        marginTop,
        marginRight,
        marginBottom,
        marginLeft,
        paddingTop,
        paddingRight,
        paddingBottom,
        paddingLeft
    };

    const renderResult = useCallback(
        resultProps => {
            const { data, error, loading } = resultProps;

            if (error) {
                return (
                    <div className={classes.error}>No products to display</div>
                );
            }

            if (loading) return '';

            if (data.products.items.length === 0) {
                return (
                    <div className={classes.error}>No products to display</div>
                );
            }

            // We have to manually resort the products
            const products = [];
            skus.forEach(sku => {
                const product = data.products.items.find(
                    product => product.sku === sku
                );
                if (product) {
                    products.push(product);
                }
            });

            return (
                <Gallery
                    items={products}
                    classes={{ items: classes.galleryItems }}
                />
            );
        },
        [classes, skus]
    );

    return (
        <div
            style={dynamicStyles}
            className={[...cssClasses, classes.root].join(' ')}
        >
            <Query query={getProductsBySku} variables={{ skus }}>
                {renderResult}
            </Query>
        </div>
    );
};

/**
 * Props for {@link Products}
 *
 * @typedef props
 *
 * @property {Object} classes An object containing the class names for the Products
 * @property {String} classes.root CSS class for products
 * @property {String} classes.galleryItems CSS class to modify child gallery items
 * @property {String} classes.error CSS class for displaying fetch errors
 * @property {Array} skus List of SKUs to load into product list
 * @property {String} textAlign Alignment of content within the products list
 * @property {String} border CSS border property
 * @property {String} borderColor CSS border color property
 * @property {String} borderWidth CSS border width property
 * @property {String} borderRadius CSS border radius property
 * @property {String} marginTop CSS margin top property
 * @property {String} marginRight CSS margin right property
 * @property {String} marginBottom CSS margin bottom property
 * @property {String} marginLeft CSS margin left property
 * @property {String} paddingTop CSS padding top property
 * @property {String} paddingRight CSS padding right property
 * @property {String} paddingBottom CSS padding bottom property
 * @property {String} paddingLeft CSS padding left property
 * @property {Array} cssClasses List of CSS classes to be applied to the component
 */
Products.propTypes = {
    classes: shape({
        root: string,
        galleryItems: string,
        error: string
    }),
    skus: arrayOf(string),
    textAlign: string,
    border: string,
    borderColor: string,
    borderWidth: string,
    borderRadius: string,
    marginTop: string,
    marginRight: string,
    marginBottom: string,
    marginLeft: string,
    paddingTop: string,
    paddingRight: string,
    paddingBottom: string,
    paddingLeft: string,
    cssClasses: arrayOf(string)
};

export default Products;
