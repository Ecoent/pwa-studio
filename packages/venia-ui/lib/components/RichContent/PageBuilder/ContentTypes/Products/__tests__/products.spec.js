import { createTestInstance } from '@magento/peregrine';
import React from 'react';
import Products from '../products';

jest.mock('@magento/venia-drivers', () => ({
    Query: jest.fn()
}));
import { Query } from '@magento/venia-drivers';
jest.mock('../../../../../Gallery', () => jest.fn());
import Gallery from '../../../../../Gallery';
const mockGallery = Gallery.mockImplementation(() => 'Gallery');

test('render products with no props & no products', () => {
    Query.mockImplementation(({ children }) => {
        return children({
            data: {
                products: {
                    items: []
                }
            },
            error: false,
            loading: false
        });
    });

    const component = createTestInstance(<Products />);

    expect(component.toJSON()).toMatchSnapshot();
});

test('render products with loading state', () => {
    Query.mockImplementation(({ children }) => {
        return children({
            data: {
                products: {
                    items: []
                }
            },
            error: false,
            loading: true
        });
    });

    const component = createTestInstance(<Products />);

    expect(component.toJSON()).toMatchSnapshot();
});

test('render products with error state', () => {
    Query.mockImplementation(({ children }) => {
        return children({
            data: {
                products: {
                    items: []
                }
            },
            error: true,
            loading: false
        });
    });

    const component = createTestInstance(<Products />);

    expect(component.toJSON()).toMatchSnapshot();
});

test('render products and ensure order is correct passed to Gallery', () => {
    Query.mockImplementation(({ children }) => {
        return children({
            data: {
                products: {
                    items: [
                        {
                            sku: 'TEST-2'
                        },
                        {
                            sku: 'TEST-1'
                        }
                    ]
                }
            },
            error: false,
            loading: false
        });
    });

    const productProps = {
        skus: ['TEST-1', 'TEST-2']
    };

    createTestInstance(<Products {...productProps} />);
    expect(mockGallery).toHaveBeenCalledWith(
        expect.objectContaining({
            items: [
                {
                    sku: 'TEST-1'
                },
                {
                    sku: 'TEST-2'
                }
            ]
        }),
        expect.anything()
    );
});
