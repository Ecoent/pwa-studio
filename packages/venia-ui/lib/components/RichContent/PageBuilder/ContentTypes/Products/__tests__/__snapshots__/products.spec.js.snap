// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render products with error state 1`] = `
<div
  className=""
  style={
    Object {
      "border": undefined,
      "borderColor": undefined,
      "borderRadius": undefined,
      "borderWidth": undefined,
      "marginBottom": undefined,
      "marginLeft": undefined,
      "marginRight": undefined,
      "marginTop": undefined,
      "paddingBottom": undefined,
      "paddingLeft": undefined,
      "paddingRight": undefined,
      "paddingTop": undefined,
      "textAlign": undefined,
    }
  }
>
  <div>
    No products to display
  </div>
</div>
`;

exports[`render products with loading state 1`] = `
<div
  className=""
  style={
    Object {
      "border": undefined,
      "borderColor": undefined,
      "borderRadius": undefined,
      "borderWidth": undefined,
      "marginBottom": undefined,
      "marginLeft": undefined,
      "marginRight": undefined,
      "marginTop": undefined,
      "paddingBottom": undefined,
      "paddingLeft": undefined,
      "paddingRight": undefined,
      "paddingTop": undefined,
      "textAlign": undefined,
    }
  }
>
  
</div>
`;

exports[`render products with no props & no products 1`] = `
<div
  className=""
  style={
    Object {
      "border": undefined,
      "borderColor": undefined,
      "borderRadius": undefined,
      "borderWidth": undefined,
      "marginBottom": undefined,
      "marginLeft": undefined,
      "marginRight": undefined,
      "marginTop": undefined,
      "paddingBottom": undefined,
      "paddingLeft": undefined,
      "paddingRight": undefined,
      "paddingTop": undefined,
      "textAlign": undefined,
    }
  }
>
  <div>
    No products to display
  </div>
</div>
`;
